trunk/src/app/srs_app_rtsp.cpp

```diff
@@ -0,0 +1,593 @@
+        if ((err = rtsp_->recv_message(&req)) != srs_success) {
+            return srs_error_wrap(err, "recv message");
+        }
+        SrsUniquePtr<SrsRtspRequest> req_ptr(req);
+        
+        if (req->is_options()) {
+            SrsUniquePtr<SrsRtspOptionsResponse> res(new SrsRtspOptionsResponse((int)req->seq));
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {
+                return  srs_error_wrap(err, "response option");
+            }
+        } else if (req->is_describe()) {
+            // create session.
+            if (session_id_.empty()) {
+                session_id_ = srs_random_str(8);
+            }
+
+            SrsUniquePtr<SrsRtspDescribeResponse> res(new SrsRtspDescribeResponse((int)req->seq));
+            res->session = session_id_;
+
+            std::string sdp;
+            err = session_->do_describe(req, sdp);
+            if (err != srs_success) {
+                res->status = SRS_CONSTS_RTSP_InternalServerError;
+                if (srs_error_code(err) == ERROR_SYSTEM_SECURITY_DENY) {
+                    res->status = SRS_CONSTS_RTSP_Forbidden;
+                }
+                srs_warn("describe failed: %s", srs_error_desc(err).c_str());
+                srs_error_reset(err);
+            }
+
+            res->sdp = sdp;
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {
+                return  srs_error_wrap(err, "response describe");
+            }
+        } else if (req->is_setup()) {
+            srs_assert(req->transport);            
+
+            SrsUniquePtr<SrsRtspSetupResponse> res(new SrsRtspSetupResponse((int)req->seq));
+            res->session = session_id_;
+
+            uint32_t ssrc = 0;
+            err = session_->do_setup(req, &ssrc);
+            if (err != srs_success) {
+                res->status = SRS_CONSTS_RTSP_InternalServerError;
+                srs_warn("setup failed: %s", srs_error_desc(err).c_str());
+                srs_error_reset(err);
+            }
+    
+            res->transport->copy(req->transport);
+            res->session = session_id_;
+            res->ssrc = srs_int2str(ssrc);
+            res->client_port_min = req->transport->client_port_min;
+            res->client_port_max = req->transport->client_port_max;
+            // TODO: FIXME: listen local port
+            res->local_port_min = 0;
+            res->local_port_max = 0;
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {  
+                return srs_error_wrap(err, "response setup");
+            }
+        } else if (req->is_play()) {
+            SrsUniquePtr<SrsRtspResponse> res(new SrsRtspResponse((int)req->seq));
+            res->session = session_id_;
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {
+                return srs_error_wrap(err, "response record");
+            }
+            err = session_->do_play(req, new SrsRtcPlayStream(this, cid_));
```

Comment:
`SrsRtcPlayStream`'s whole lifecycle can be totally managed inside `SrsRtspSession`, construct instance when `do_play`, destruct when `do_teardown()` and `session_`'s destruction.

```suggestion
            err = session_->do_play(req, this, cid_);
            
           virtual srs_error_t do_play(SrsRtspRequest* req, SrsRtcConnction* conn, const SrsContextId& cid);
```

Reply:
嗯，有道理